### ISSUE TITLE
Avoid two step process for adding CI to a repo

### ISSUE THREAD

**Issue by aidanheerdegen** [ACCESS-NRI/build-ci#14]
Currently adding CI to a repo requires two steps, adding the package to build-ci to get dependencies in the base image and adding the workflow to the repo, e.g.

https://github.com/ACCESS-NRI/cice5/issues/4

https://github.com/ACCESS-NRI/build-ci/issues/13

Is there a way we can avoid this, or streamline in some way?

**Comment by aidanheerdegen** [comment]
Any ideas @echus @CodeGat @Quartzer2 ?

**Comment by echus** [comment]
Yeah, I think we had a discussion with @Quartzer2 @harshula about whether building separate images is actually necessary. It might make sense in our case to generate a single base image containing all the dependencies required to build *all* the models we support, in which case this issue goes away. It would help to know:

- How many models we are expecting to support
- How much their dependencies overlap
- Whether we'll need to rapidly add support for new ones on a regular basis.

**Comment by CodeGat** [comment]
Just to check my understanding, would we be looking to have a single image with all the relevant spack environments (compilers etc etc) and model repositories as a base image? Wouldn't this lead to the creation of many variants of this one large image whenever there are merges to any model?
Could we house the model-specific `workflow.yml`s in the build-ci repo instead of the model repo? Is there any information that is in the model repos that can't be accessed from the build-ci repo? Can we kick off a build-ci workflow upon pushes to other repositories?

**Comment by CodeGat** [comment]
After getting a bit more familiar with the codebase, I've been looking into the 'single base image' stuff that @echus was talking about. 
For an image that contains all the models, could we utilize `spack env`s? Could each model have an associated environment that contains all the dependencies within the image, which is activated at the end (`ENTRYPOINT spack env activate $PACKAGE_ENV` maybe?)
I can take a look at this at the moment.

**Comment by CodeGat** [comment]
I've been testing building multiple models (and their dependencies) via different spack environments in the same image, and here are my results so far (note the sizes)
```
spackenv-test (mom5, cice5 and access-om3)    v3        d20f040c9ce4   6 seconds ago    7.63GB
spackenv-test (mom5 and cice5)                v2        393ce2874a30   17 minutes ago   7.28GB
spackenv-test (just mom5)                     v1        e8b68ecdac38   27 minutes ago   7.27GB
ghcr.io/access-nri/build-mom5-intel2021.1.2   latest    42d6e1b15a57   2 weeks ago      5.78GB
spack/rockylinux8                             v0.20.1   51847f373f65   2 weeks ago      958MB
```
spackenv-test s are based on spack/rockylinux8 .
Some things:
* spackenv-test:v1 is 2GB larger than our equivalent  build-mom5-intel2021.1.2 (however I believe this is because I was docker committing between installs of models to save the state, which is far more space-inefficient than just making a dockerfile with the commands that need to be run - I suspect that the final image will be closer to build-mom5-intel2021.1.2)
* The differences between v1, v2 and v3 are relatively minimal - only 400MB. I think that we can get away with having a single image with all the dependencies for all the models in spack envs. If build caching works, building an image with an updated model shouldn't be an issue, unless we're updating a really low-level dependency.
* If we go with a single image, we would also be saving space for more github-hosted packages - does anyone know if we have a limit?
* This could also simplify the docker workflows in build-ci

**Comment by CodeGat** [comment]
We could add a very basic, generic workflow that calls a more detailed build-ci workflow that can note the compiler version etc etc. and make that repository a template repo for all future models. There would not be any dev overhead.

**Comment by CodeGat** [comment]
This was actually closed as a result of #91 ! Hooray!
